### PR TITLE
add form to "invoices/:id"

### PR DESCRIPTION
**Pull request by manandearth**
## Goal
- "/invoices/:id" to show a form same as "/invoices-insert" that shows the fields filled with the actual data in order to update or cancel.
- Both `invoices-insert` and `invoices-update` will return a table of the updated db.

## Proposed changes
- DONE add update endpoint and logic. make a view.                                                                                                 
- DONE need to spec what the amount field in the form.                                                                                             
- DONE submitting the form should return "/invoices"                                                                                               
- DONE split views to the different services                                                                                                            
- DONE In invoices-update, pass :id as path-param :amount as form-param  
>refactor the "/invoices-update" route into "/invoices-update/:id". Then you get the id from :path-params, and the email name from :form-params.                                                                               
- DONE call the views from the handler :
>  The **handler** invoking the perform function that returns something like:
{:status 200 :body (views/foo)}

## PR status, roadblocks, etc

`submit-invoice` function repeated in two views should it be in the more general views.clj instead?

### CODE REVIEW

**Comment by manandearth** [comment]
I'll get to it later, nothing to review at the moment.

File: src/playground/services/invoices/update/endpoint.clj [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+   [clojure.java.jdbc :as jdbc]
+   [clojure.spec.alpha :as spec]
+   [honeysql.core :as h]
+   [playground.services.invoices.update.logic :as logic]))
+
+(spec/def ::amount nat-int?)
+
+(spec/def ::api (spec/keys :req-un [::amount]))
+
+(defn perform [{{:keys [id amount]} :form-params :keys [db] :as request}]
```
Comment (manandearth) [inline src/playground/services/invoices/update/endpoint.clj]: basically copied from invoices-insert, so not entirely all useful.

File: src/playground/services/invoices/update/logic.clj [hunk 2]
```diff
@@ -0,0 +1,20 @@
+(ns playground.services.invoices.update.logic
+  (:require
+   [honeysql.core :as h]
+   [honeysql.helpers :as hh]
+   [honeysql.types :as ht]
```
Comment (manandearth) [inline src/playground/services/invoices/update/logic.clj]: ended up not using [honeysql.types..]

**Comment by vemv** [comment]
LMK when the pr is ready

Update its description if needed

File: .github/pull_request_template.md [hunk 3]
```diff
@@ -2,6 +2,11 @@
...
 
 ## Proposed changes
-add update endpoint and logic. make a view.
+- add update endpoint and logic. make a view.
+- need to spec what the amount field in the form.
+- submitting the form should return "/invoices"
 
 ## PR status, roadblocks, etc
+ - :id and :amount pass as form-params in "/invoices-update" and can be updated in the form yet id shouldn't be update as it is a SERIAL in the DB.
+
```
Comment (manandearth) [inline .github/pull_request_template.md:7]: the reason the :id is in the form is to carry it over to the "UPDATE users SET amount ? WHERE id ,,,"
Need a different way to carry :id not as form-param

**Comment by manandearth** [comment]
I'll get back to it in the morning

File: src/playground/services/invoices/update/endpoint.clj [hunk 4]
```diff
@@ -13,10 +14,8 @@
...
+        fetch  (h/format (retrieve-all.logic/query-all))
         _      (jdbc/execute! db update)
-        result (-> (jdbc/query db fetch)
-                   (logic/to-serialize))]
-
-    {:status 200
-     :body   "OK"}))
+        result (jdbc/query db fetch)]
+    
+    {:result result :confirmation (str "The entry " id " has been updated.")}))
```
Comment (manandearth) [inline src/playground/services/invoices/update/endpoint.clj]: what do you think of returning such a map? is it too obscure?

File: src/playground/views.clj [hunk 5]
```diff
@@ -67,6 +68,24 @@
        [:p [:label "amount: " [:input {:type "text" :name "amount" :value (first (clojure.string/split (:email user) #"@"))}]]]
        [:p [:label "λ ->" [:input {:type "submit" :value "Update"}]]]]]]))
 
+(defn update-submit-invoice [context]
```
Comment (manandearth) [inline src/playground/views.clj]: Here is where that map is being used .

**Comment by vemv** [comment]
Is the PR ready?

**Comment by manandearth** [comment]
I wanted you to review the invoices-update and what I commented in the code, and if it's a good direction I'll create a similar route/view for invoices-insert. then the pr will be ready.
EDIT:
I'll just go ahead and prepare the other route anyway...

**Comment by manandearth** [comment]
Ready for review.
BTW, Do I use "context" properly here?

File: .github/pull_request_template.md [hunk 6]
```diff
@@ -1,5 +1,7 @@
 ## Goal
+"/invoices/:id" to show a form same as "/invoices-insert" that shows the fields filled with the actual data in order to update or cancel.
```
Comment (vemv) [inline .github/pull_request_template.md]: Revert these changes

Templates are _templates_. Each time you create a new PR, the template will be applied, and you will complete it on-the-fly.

i.e. you don't want this specific info to show up in new unrelated PRs.

**Comment by vemv** [comment]
Fixed the endpoint, please study the commit

The `amount` param (and spec) didn't make sense. If you aren't dealing with amounts, don't name things _amount_ and also don't expect them to be integers (email _names_ aren't integers)

File: .github/pull_request_template.md (hunk 3 above)
Comment (vemv) [reply .github/pull_request_template.md:7]: Gotchu. The proper way would be to refactor the `"/invoices-update"` route into `"/invoices-update/:id"`. Then you get the id from :path-params, and the email name from :form-params.

Please do give it a try

File: src/playground/views.clj [hunk 7]
```diff
@@ -3,7 +3,8 @@
    [hiccup.page :as page]
    [hiccup.table :as table]
    [playground.services.invoices.retrieve-all.endpoint :as retrieve-all]
-   [playground.services.invoices.retrieve.endpoint :as retrieve]))
+   [playground.services.invoices.retrieve.endpoint :as retrieve]
+   [playground.services.invoices.update.endpoint :as invoices.update]))
```
Comment (vemv) [inline src/playground/views.clj]: This require serves no purpose?

Delete them along the similar ones above

File: src/playground/services/invoices/update/endpoint.clj (hunk 4 above)
Comment (vemv) [reply src/playground/services/invoices/update/endpoint.clj]: Not clean yep.

Will leave another comment with the answer

File: src/playground/service.clj [hunk 8]
```diff
@@ -25,15 +26,21 @@
...
-(defn invoice-page [request]
+(defn update-page [request]
   (let [user (invoices.retrieve/perform request)]
     (if user
-      (ring-resp/response (views/invoice user))
+      (ring-resp/response (views/update-invoice user))
       (ring-resp/not-found "Entry not in DB"))))
 
+(defn update-submit-page [request]
+  (ring-resp/response (views/submit-invoice (invoices.update/perform request))))
```
Comment (vemv) [inline src/playground/service.clj]: This pattern is wrong. Handlers should be the responsible for invoking views, not the opposite.

Please fix it throughout this ns

File: src/playground/views.clj [hunk 9]
```diff
@@ -3,7 +3,8 @@
    [hiccup.page :as page]
    [hiccup.table :as table]
    [playground.services.invoices.retrieve-all.endpoint :as retrieve-all]
-   [playground.services.invoices.retrieve.endpoint :as retrieve]))
+   [playground.services.invoices.retrieve.endpoint :as retrieve]
```
Comment (vemv) [inline src/playground/views.clj]: Once you address all other feedback, and you are sure that everything works correctly, as a last commit split this ns across the "services".

Each service will have the following structure:

```
.
├── endpoint.clj
├── logic.clj
└── views.clj
```

File: src/playground/service.clj (hunk 8 above)
Comment (manandearth) [reply src/playground/service.clj]: Do I then understand correctly that the pattern should be :

The route invoking the `perform` function that returns something like:
`{:status 200 :body (views/foo)}`

Comment (vemv) [reply src/playground/service.clj]: Yes :)

> The route invoking the perform function 

We can call these 'handlers'

File: .github/pull_request_template.md (hunk 6 above)
Comment (manandearth) [reply .github/pull_request_template.md]: Sorry V, I don't quite get it. If not on the template file, where does this md exist for me to update?

Comment (manandearth) [reply .github/pull_request_template.md]: I think I get it... I edit the pr description at the top on github.

**Comment by manandearth** [comment]
merge this and continue with `spec` and a delete route?

**Comment by vemv** [comment]
>  and continue with spec

what spec?

File: .github/pull_request_template.md (hunk 6 above)
Comment (vemv) [reply .github/pull_request_template.md]: Yes, `.github/pull_request_template.md` remains generic and half-empty and you edit things on github

**Comment by manandearth** [comment]
I'll have to look but I don't think spec inteceptor is everywhere needed.

File: src/playground/services/invoices/update/logic.clj [hunk 10]
```diff
@@ -0,0 +1,18 @@
+(ns playground.services.invoices.update.logic
+  (:require
+   [honeysql.core :as h]
+   [honeysql.helpers :as hh]))
+
+(defn to-update [id name uuid]
+  (-> (hh/update :users)
+      (hh/sset  {:email (str name "@" uuid ".com")})
+      (hh/where [:= :id (Integer/parseInt id)])))
```
Comment (vemv) [inline src/playground/services/invoices/update/logic.clj]: Do the parsing in the handler.

In an MVC architecture, (C being the handler, and M being the logic), validating/coercing user input is responsbility of the C.

If possible, do it via the spec interceptor rather than `Integer/parseInt`

**Review by vemv — COMMENTED** [review]
Better fix the spec aspect as part of this PR, then we're done!

**Comment by manandearth** [comment]
The spec I mention , I see now is the new spec `string?` that replaced the `int?`
in update ns: https://github.com/manandearth/playground/blob/bf5d511ffd996090750e60ef1781a21c32b314c5/src/playground/services/invoices/update/endpoint.clj#L11

It means that updating now can be any string. I don't mind for this example that it will stay.

**Comment by manandearth** [comment]
I don't see it as a spec issue as we have a text user-input which needs parsing to an int to match SERIAL :id in DB.
Am I missing something?

**Comment by vemv** [comment]
If you spec the id as an integer, the "coercer" interceptor will transform the id param to int.

Which saves you the `(Integer/parseInt)` call, which is repetitive and error-prone in larger apps

**Comment by vemv** [comment]
You can give it a try, OTOH it may fail because spec-coercion only worked fine over `:form-params` but not over the broader `:params`, last time we tried.

**Comment by manandearth** [comment]
I tried:
```
 ["/invoices-update/:id" :post (into common-interceptors [http/json-body (param-spec-interceptor ::invoices.update/api [:form-params :path-p\
   arams]) `invoices.update/perform])]
```

and:
```
 ["/invoices-update/:id" :post (into common-interceptors [http/json-body (param-spec-interceptor ::invoices.update/api :form-params ) (param-spec-interceptor ::invoices.update/api :path-params ) `invoices.update/perform])]
```
doesn't cut the cheese..

**Comment by manandearth** [comment]
```
{"explanation":"{:id \"33\", :playground.coerce/invalid? true} - failed: nat-int? spec: :playground.services.invoices.update.endpoint/id\n"}
```

**Comment by vemv** [comment]
I'll give it a shot soonish. Feel free to move on, merge and new branch / PR

Good work, genuinely thrilled to see the progess!
